Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -287,13 +288,21 @@ void NetworkStorageManager::close(CompletionHandler<void()>&& completionHandler)
     });
 }
 
-void NetworkStorageManager::startReceivingMessageFromConnection(IPC::Connection& connection, const Vector<WebCore::RegistrableDomain>& allowedSites)
+void NetworkStorageManager::startReceivingMessageFromConnection(IPC::Connection& connection, const Vector<WebCore::RegistrableDomain>& allowedSites, const SharedPreferencesForWebProcess& preferences)
 {
     ASSERT(RunLoop::isMain());
 
     connection.addWorkQueueMessageReceiver(Messages::NetworkStorageManager::messageReceiverName(), m_queue.get(), *this);
     m_connections.add(connection);
     addAllowedSitesForConnection(connection.uniqueID(), allowedSites);
+
+    protectedWorkQueue()->dispatch([this, protectedThis = Ref { *this }, connection = connection.uniqueID(), preferences]() mutable {
+        assertIsCurrent(workQueue());
+        ASSERT(!m_preferencesForConnections.contains(connection));
+        m_preferencesForConnections.add(connection, preferences);
+
+        RunLoop::protectedMain()->dispatch([protectedThis = WTFMove(protectedThis)] { });
```

Comment:
It is a bit sad we have to do this, maybe we should refactor the code a bit so that we can make sure NetworkStorageManager is DestructionThread::MainRunLoop.
We can probably do this by making WorkQueueMessageReceiver AbstractRefCounted and adding a ThreadSafe WorkQueueMessageReceiver or something like that.

Reply:
Good idea, I filed https://bugs.webkit.org/show_bug.cgi?id=284065.